modules/twinklefluff.js

```diff
@@ -379,121 +381,61 @@ Twinkle.fluff.revert = function revertPage(type, vandal, rev, page) {
-			revertToUserHidden ? null : revertToUser, optional_summary);
-
-		var query = {
-			action: 'edit',
-			title: mw.config.get('wgPageName'),
-			summary: summary,
-			tags: Twinkle.changeTags,
-			token: csrftoken,
-			undo: lastrevid,
-			undoafter: revertToRevID,
-			basetimestamp: touched,
-			starttimestamp: loadtimestamp,
-			minor: Twinkle.getPref('markRevertedPagesAsMinor').indexOf('torev') !== -1 ? true : undefined,
-			format: 'json'
-		};
-		// Handle watching, possible expiry
-		if (Twinkle.getPref('watchRevertedPages').indexOf('torev') !== -1) {
-			var watchOrExpiry = Twinkle.getPref('watchRevertedExpiry');
+		var summary = Twinkle.fluff.formatSummary('Restored revision ' + pageobj.getRevisionID() + ' by $USER',
+			pageobj.getRevisionUser(), optional_summary);
 
-			if (!watchOrExpiry || watchOrExpiry === 'no') {
-				query.watchlist = 'nochange';
-			} else if (watchOrExpiry === 'default' || watchOrExpiry === 'preferences') {
-				query.watchlist = 'preferences';
-			} else {
-				query.watchlist = 'watch';
-				// number allowed but not used in Twinkle.config.watchlistEnums
-				if ((!page.watched || page.watchlistexpiry) && typeof watchOrExpiry === 'string' && watchOrExpiry !== 'yes') {
-					query.watchlistexpiry = watchOrExpiry;
-				}
-			}
+		pageobj.setChangeTags(Twinkle.changeTags);
+		pageobj.setEditSummary(summary);
+		if (Twinkle.getPref('watchRevertedPages').indexOf('torev') !== -1) {
+			pageobj.setWatchlist(Twinkle.getPref('watchRevertedExpiry'));
+		}
+		if (Twinkle.getPref('markRevertedPagesAsMinor').indexOf('torev') !== -1) {
+			pageobj.setMinorEdit(true);
 		}
 
-		Morebits.wiki.actionCompleted.redirect = mw.config.get('wgPageName');
+		Morebits.wiki.actionCompleted.redirect = pageobj.getPageName();
 		Morebits.wiki.actionCompleted.notice = 'Reversion completed';
 
-		var wikipedia_api = new Morebits.wiki.api('Saving reverted contents', query, Twinkle.fluff.callbacks.complete, apiobj.statelem);
-		wikipedia_api.params = apiobj.params;
-		wikipedia_api.post();
+		pageobj.revert();
 	},
 	main: function(apiobj) {
-		var response = apiobj.getResponse();
-
-		var loadtimestamp = response.curtimestamp;
-		var csrftoken = response.query.tokens.csrftoken;
-
-		var page = response.query.pages[0];
+		var page = apiobj.getResponse().query.pages[0];
 		if (!page.actions.edit) {
 			apiobj.statelem.error("Unable to edit the page, it's probably protected.");
 			return;
 		}
 
-		var lastrevid = parseInt(page.lastrevid, 10);
-		var touched = page.touched;
-
-		var revs = page.revisions;
-
 		var statelem = apiobj.statelem;
 		var params = apiobj.params;
 
+		var lastrevid = parseInt(page.lastrevid, 10);
+		var revs = page.revisions;
 		if (revs.length < 1) {
 			statelem.error('We have less than one additional revision, thus impossible to revert.');
 			return;
 		}
 		var top = revs[0];
 		var lastuser = top.user;
 
+		// Should be handled by the API, but nice to quit early
 		if (lastrevid < params.revid) {
```

Comment:
The API handles this, but we could do a similar early-fail in Morebits if we wanted.